lib/march_hare/session.rb

```diff
@@ -507,6 +507,7 @@ def new_connection_impl(hosts, host_selector)
 
       converting_rjc_exceptions_to_ruby do
         if @executor_factory
+          @executor.shutdown() if @executor
```

Comment:
I suggest using [awaitTermination](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorService.html#awaitTermination(long,%20java.util.concurrent.TimeUnit)) with a timeout (say, of 30 seconds by default).